File: 프로그래머스/unrated/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> time_split(string date) {
    int year = 0, s = 1000;
    for (int i = 0; i < 4; i++) {
        year += (date[i] - '0') * s;
        s /= 10;
    }
    int month = (date[5] - '0') * 10 + (date[6] - '0');
    int day = (date[8] - '0') * 10 + (date[9] - '0');
    return { year, month, day };
}

vector<int> date_plus(vector<int> ymd, int p) {
    int y = ymd[0], m = ymd[1], d = ymd[2];
    d--;
    if (!d)  d = 28, m--;
    m += p;
    y += (m - 1) / 12;
    m %= 12;
    if (!m)  m = 12;
    return { y,m,d };
}

vector<int> solution(string today, vector<string> terms, vector<string> privacies) {
    vector<int> answer;
    vector<int> real_today = time_split(today);
    int real_terms[26]{};
    for (string i : terms) {
        char t = i[0] - 'A';
        int period = 0, s = 1;
        for (int j = i.size() - 1; j >= 2; j--) {
            period += (i[j] - '0') * s;
            s *= 10;
        }
        real_terms[t] = period;
    }
    int id = 1;
    for (string i : privacies) {
        string parsed = "";
        char term = ' ';
        bool passed = false;
        for (char j : i) {
            if (j == ' ') {
                passed = true;
                continue;
            }
            if (passed)  term = j;
            else    parsed += j;
        }
        vector<int> ymd = date_plus(time_split(parsed), real_terms[term - 'A']);
        
        if (real_today[0] > ymd[0])  answer.push_back(id);
        else if (real_today[0] == ymd[0]) {
            if (real_today[1] > ymd[1])  answer.push_back(id);
            else if (real_today[1] == ymd[1]) {
                if (real_today[2] > ymd[2])  answer.push_back(id);
            }
        }

        id++;
    }

    return answer;
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.util.*;

class Solution {
    public int dateToInteger(String date) {
        String[] sp = date.split("\\.");
        int year = Integer.parseInt(sp[0]);
        int month = Integer.parseInt(sp[1]);
        int day = Integer.parseInt(sp[2]);
        
        int ret = (day-1) + (month-1) * 28 + year * 28 * 12;
        return ret;
    }
    
    public int[] solution(String today, String[] terms, String[] privacies) {
        int todayValue = dateToInteger(today);
        
        int[] limit = new int[128];
        for(String term : terms) {
            String[] sp = term.split(" ");
            char type = sp[0].charAt(0);
            limit[type] = Integer.parseInt(sp[1]);
        }
        
        List<Integer> list = new ArrayList<>();
        for(int i=0;i<privacies.length;i++) {
            String privacy = privacies[i];
            
            String[] sp = privacy.split(" ");
            String date = sp[0];
            char type = sp[1].charAt(0);
            
            int privacyDateValue = dateToInteger(date);
            if(todayValue >= privacyDateValue + limit[type] * 28) {
                list.add(i+1);
            }
        }
        
        int[] answer = new int[list.size()];
        for(int i=0;i<list.size();i++) {
            answer[i] = list.get(i);
        }
        
        return answer;
    }
}